deepmd/pt/utils/env_mat_stat.py

```diff
@@ -0,0 +1,205 @@
+
+        Parameters
+        ----------
+        env_mat : torch.Tensor
+            The environment matrix.
+
+        Returns
+        -------
+        Dict[str, StatItem]
+            The statistics of the environment matrix.
+        """
+        stats = {}
+        for kk, vv in env_mat.items():
+            stats[kk] = StatItem(
+                number=vv.numel(),
+                sum=vv.sum().item(),
+                squared_sum=torch.square(vv).sum().item(),
+            )
+        return stats
+
+
+class EnvMatStatSeA(EnvMatStat):
+    """Environmental matrix statistics for the se_a environemntal matrix.
+
+    Parameters
+    ----------
+    descriptor : DescriptorBlock
+        The descriptor of the model.
+    """
+
+    def __init__(self, descriptor: "DescriptorBlock"):
+        super().__init__()
+        self.descriptor = descriptor
+
+    def iter(
+        self, data: List[Dict[str, torch.Tensor]]
+    ) -> Iterator[Dict[str, StatItem]]:
+        """Get the iterator of the environment matrix.
+
+        Parameters
+        ----------
+        data : List[Dict[str, torch.Tensor]]
+            The environment matrix.
+
+        Yields
+        ------
+        Dict[str, StatItem]
+            The statistics of the environment matrix.
+        """
+        zero_mean = torch.zeros(
+            self.descriptor.get_ntypes(),
+            self.descriptor.get_nsel(),
+            4,
+            dtype=env.GLOBAL_PT_FLOAT_PRECISION,
+            device=env.DEVICE,
+        )
+        one_stddev = torch.ones(
+            self.descriptor.get_ntypes(),
+            self.descriptor.get_nsel(),
+            4,
+            dtype=env.GLOBAL_PT_FLOAT_PRECISION,
+            device=env.DEVICE,
+        )
+        for system in data:
+            coord, atype, box, natoms = (
+                system["coord"],
+                system["atype"],
+                system["box"],
+                system["natoms"],
+            )
+            (
+                extended_coord,
+                extended_atype,
+                mapping,
+                nlist,
+            ) = extend_input_and_build_neighbor_list(
+                coord,
+                atype,
+                self.descriptor.get_rcut(),
+                self.descriptor.get_sel(),
+                distinguish_types=self.descriptor.distinguish_types(),
+                box=box,
+            )
+            env_mat, _, _ = prod_env_mat_se_a(
+                extended_coord,
+                nlist,
+                atype,
+                zero_mean,
+                one_stddev,
+                self.descriptor.get_rcut(),
+                # TODO: export rcut_smth from DescriptorBlock
+                self.descriptor.rcut_smth,
+            )
+            env_mat = env_mat.view(
+                coord.shape[0], coord.shape[1], self.descriptor.get_nsel(), 4
+            )
+            env_mats = {}
+
+            if "real_natoms_vec" not in system:
```

Comment:
shouldn't be determined by the descriptor's attribute `distinguish_type`?

Reply:
No. To my understanding, `mixed_type` (produce `real_natoms_vec` here) describes the central atom i from the data, while `distinguish_type` describes the neighboring atoms. Here, the central atoms matter.

Reply:
No. To my understanding, mixed_type (produce real_natoms_vec here) describes the central atom i from the data, while distinguish_type describes the neighboring atoms from the descriptor. Here, the central atoms matter.